src/dbnode/storage/bootstrap/bootstrapper/persist.go

```diff
@@ -112,97 +117,96 @@ func PersistBootstrapIndexSegment(
+		var (
+			fulfilled         = indexBlock.Fulfilled()
+			persistedSegments []segment.Segment
+		)
+		defer func() {
+			if !success {
+				return
+			}
 
-		// Now replace the active segment with the persisted segment.
-		newFulfilled := fulfilled.Copy()
-		newFulfilled.AddRanges(expectedRanges)
-		replacedBlock := result.NewIndexBlock(blockStart, segments, newFulfilled)
-		indexResults[xtime.ToUnixNano(blockStart)] = replacedBlock
-	}()
+			// Combine persisted  and existing segments.
+			segments := make([]segment.Segment, 0, len(persistedSegments))
+			for _, pSeg := range persistedSegments {
+				segments = append(segments, NewSegment(pSeg, true))
+			}
+			for _, seg := range indexBlock.Segments() {
+				segments = append(segments, seg)
+			}
 
-	// Check that we completely fulfilled all shards for the block
-	// and we didn't bootstrap any more/less than expected.
-	requireFulfilled := expectedRanges.Copy()
-	requireFulfilled.Subtract(fulfilled)
-	exactStartEnd := max.Equal(blockStart.Add(blockSize))
-	if !exactStartEnd || !requireFulfilled.IsEmpty() {
-		return fmt.Errorf("persistent fs index bootstrap invalid ranges to persist: "+
-			"expected=%v, actual=%v, fulfilled=%v, exactStartEnd=%v, requireFulfilledEmpty=%v",
-			expectedRanges.String(), requestedRanges.String(), fulfilled.String(),
-			exactStartEnd, requireFulfilled.IsEmpty())
-	}
+			// Now replace the active segment with the persisted segment.
+			newFulfilled := fulfilled.Copy()
+			newFulfilled.AddRanges(expectedRanges)
+			replacedBlock := result.NewIndexBlock(segments, newFulfilled)
+			indexResults[xtime.ToUnixNano(blockStart)].Data[volumeType] = replacedBlock
+		}()
 
-	// NB(r): Need to get an exclusive lock to actually write the segment out
-	// due to needing to incrementing the index file set volume index and also
-	// using non-thread safe resources on the persist manager.
-	persistManager.Lock()
-	defer persistManager.Unlock()
+		// Check that we completely fulfilled all shards for the block
+		// and we didn't bootstrap any more/less than expected.
+		requireFulfilled := expectedRanges.Copy()
+		requireFulfilled.Subtract(fulfilled)
+		exactStartEnd := max.Equal(blockStart.Add(blockSize))
+		if !exactStartEnd || !requireFulfilled.IsEmpty() {
+			return fmt.Errorf("persistent fs index bootstrap invalid ranges to persist: "+
+				"expected=%v, actual=%v, fulfilled=%v, exactStartEnd=%v, requireFulfilledEmpty=%v",
+				expectedRanges.String(), requestedRanges.String(), fulfilled.String(),
+				exactStartEnd, requireFulfilled.IsEmpty())
+		}
 
-	flush, err := persistManager.Mgr.StartIndexPersist()
-	if err != nil {
-		return err
-	}
+		// NB(r): Need to get an exclusive lock to actually write the segment out
+		// due to needing to incrementing the index file set volume index and also
+		// using non-thread safe resources on the persist manager.
+		persistManager.Lock()
+		defer persistManager.Unlock()
```

Comment:
Hm this will double lock inside the for loop now yeah? (since second iteration of the loop will try to lock again before unlocked at the end of the function)

Maybe best to make everything inside this for loop a separate function now? (then the defer and locking all stays the same)